hub/services/hub.go

```
@@ -133,11 +136,70 @@ func (h *Hub) Start() error {
 	return err
 }
 
-func (h *Hub) Stop() {
+func (h *Hub) StopServices(ctx context.Context, in *idl.StopServicesRequest) (*idl.StopServicesReply, error) {
+	err := h.StopAgents()
+	if err != nil {
+		gplog.Debug("failed to stop agents: %#v", err)
+	}
+
+	h.Stop(false)
+	return &idl.StopServicesReply{}, nil
+}
+
+// TODO: add unit tests for this; this is currently tricky due to h.AgentConns()
+//    mutating global state
+func (h *Hub) StopAgents() error {
+	// FIXME: h.AgentConns() fails fast if a single agent isn't available
+	//    we need to connect to all available agents so we can stop just those
+	_, err := h.AgentConns()
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	errs := make(chan error, len(h.agentConns))
+
+	for _, conn := range h.agentConns {
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			_, err := conn.AgentClient.StopAgent(context.Background(), &idl.StopAgentRequest{})
+			if err != nil {
+				errCode := grpcStatus.Code(err)
+				errMsg := grpcStatus.Convert(err).Message()
+				// XXX: "transport is closing" is not documented but is needed to uniquely interpret codes.Unavailable
+				// https://github.com/grpc/grpc/blob/v1.24.0/doc/statuscodes.md
+				if errCode != codes.Unavailable || errMsg != "transport is closing" {
+					errs <- xerrors.Errorf("failed to stop agent on host %s : %w", conn.Hostname, err)
+				}
+				return
+			}
+
+			errs <- errors.Errorf("failed to stop agent on host: %s", conn.Hostname)
```

Don't we get here only when `err == nil`?

---

Yes. Funnily enough, a lack of error means we didn't really stop the agent, as long as we keep the abortive-`Stop` implementation. Maybe the error message should make that clear?

---

Since the "happy path" is the one with the transport error, maybe we should also invert this:
```go
if err == nil {
    errs <- errors.Errorf("failed to stop agent on %s (lack of expected error)", conn.Hostname)
    return
}

status := grpcStatus.Convert(err)
...
```

---

I think I find this cleaner.

---

@kalensk The inverted version or the code as is?

---

I modified the code to have the "happy path" not under a conditional.  This seems cleaner, and emphasizes that we really do intend an error return to indicate a successful call.  Note some users of `gpupgrade stop-services` might not care if some segment hosts don't have agents running, but they are free to ignore the error.

---

Late to the party here, but I meant I preferred where the happy path is not under conditional, so

```
if err == nil {
    errs <- errors.Errorf("failed to stop agent on %s (lack of expected error)", conn.Hostname)
    return
}

status := grpcStatus.Convert(err)
...
```